File: shared/src/main/java/xyz/klinker/messenger/shared/util/PromotionUtils.java
package xyz.klinker.messenger.shared.util;

import android.app.AlertDialog;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Handler;
import android.widget.Toast;

import xyz.klinker.messenger.shared.R;
import xyz.klinker.messenger.shared.activity.RateItDialog;
import xyz.klinker.messenger.shared.data.MimeType;
import xyz.klinker.messenger.shared.data.Settings;

public class PromotionUtils {

    private Context context;
    private SharedPreferences sharedPreferences;

    public PromotionUtils(Context context) {
        this.context = context;
        this.sharedPreferences = Settings.get(context).getSharedPrefs(context);
    }

    public void checkPromotions() {
        if (shouldAskForRating()) {
            askForRating();
        }
    }

    private boolean shouldAskForRating() {
        String pref = "install_time";
        long currentTime = System.currentTimeMillis();
        long installTime = sharedPreferences.getLong(pref, -1L);

        if (installTime == -1L) {
            // write the install time to now
            sharedPreferences.edit().putLong(pref, currentTime).apply();
        } else {
            if (currentTime - installTime > TimeUtils.TWO_WEEKS) {
                return sharedPreferences.getBoolean("show_rate_it", true);
            }
        }

        return false;
    }

    private void askForRating() {
        sharedPreferences.edit().putBoolean("show_rate_it", false)
                .apply();

        new Handler().postDelayed(() ->
                context.startActivity(new Intent(context, RateItDialog.class)),
                1000);
    }
}

File: shared/src/main/java/xyz/klinker/messenger/shared/util/PromotionUtils.kt
package xyz.klinker.messenger.shared.util

import android.app.Activity
import android.os.Handler
import com.sensortower.rating.RatingPrompt
import com.sensortower.rating.RatingPromptOptions
import com.sensortower.rating.RatingPromptSettings
import xyz.klinker.messenger.api.implementation.Account
import xyz.klinker.messenger.shared.data.Settings

class PromotionUtils(private val context: Activity) {

    fun checkPromotions(onTrialExpired: () -> Unit) {
        if (trialExpired()) {
            onTrialExpired()
        } else {
            askForRating()
        }
    }

    private fun trialExpired(): Boolean {
        return Account.exists() && Account.subscriptionType == Account.SubscriptionType.FREE_TRIAL && Account.getDaysLeftInTrial() <= 0
    }

    private fun askForRating() {
        if (Account.exists() && !Account.primary) {
            // only prompt for rating on the primary device
            return
        }

        val settings = RatingPromptSettings.getInstance(context)
        val sharedPrefs = Settings.getSharedPrefs(context)

        if (sharedPrefs.getBoolean("update_rating_method", true)) {
            sharedPrefs.edit().putBoolean("update_rating_method", false).commit()

            settings.putBoolean("rating-prompt-dont-show-again", false)
            settings.putLong("rating-prompt-should-show-at-timestamp", -1L)
        }

        Handler().postDelayed({
            RatingPrompt.show(context, RatingPromptOptions.Builder()
                    .useAlternateStyle(RatingPromptOptions.Popup.Builder("Pulse")
                            .accentColor(Settings.mainColorSet.color)
                            .darkTheme(Settings.isCurrentlyDarkTheme(context))
                    ).build())
        }, 500)
    }
    
}
